cardano-testnet/src/Testnet/Components/Query.hs

```diff
@@ -98,25 +108,117 @@ waitForEpochs epochStateView@EpochStateView{nodeConfigPath, socketPath} interval
   currentEpoch <- getCurrentEpochNo epochStateView
   waitUntilEpoch nodeConfigPath socketPath $ addEpochInterval currentEpoch interval
 
+
+-- | Wait for the requested number of blocks
+waitForBlocks
+  :: HasCallStack
+  => MonadIO m
+  => MonadTest m
+  => MonadAssertion m
+  => EpochStateView
+  -> Word64 -- ^ Number of blocks to wait
+  -> m BlockNo -- ^ The block number reached
+waitForBlocks epochStateView numberOfBlocks = withFrozenCallStack $ do
+  BlockNo startingBlockNumber <- getBlockNumber epochStateView
+  fmap BlockNo $
+    watchEpochStateUpdate epochStateView $ \(_, _, BlockNo blockNumber) ->
+      pure $
+        if blockNumber >= startingBlockNumber + numberOfBlocks
+        then Just blockNumber
+        else Nothing
+
 -- | A read-only mutable pointer to an epoch state, updated automatically
 data EpochStateView = EpochStateView
   { nodeConfigPath :: !(NodeConfigFile In)
+  -- ^ node configuration file path
   , socketPath :: !SocketPath
-  , epochStateView :: !(IORef (Maybe AnyNewEpochState))
+  -- ^ node socket path, to which foldEpochState is connected to
+  , wakeLock :: !(TChan ())
```

Comment:
What was wrong with the `IORef`? If we are going to use `TChan`s and `TMVar`s there needs to be good justification for doing so. I don't immediately see the benefit of this increased complexity in an already tedious testing environment.

Reply:
The goal here is to listen to changes of `EpochStateView`. You can't do this just using `IORef`. You could do this using `MVar`, and having watching threads block on empty `MVar` and continue when `MVar` gets filled. But then you still need additional `IORef` for functions checking for current `EpochStateView` without blocking. The problem with using `MVar` and `IORef` at the same time, is that you don't have any guarantees about [the order of operations in the listening thread](https://hackage.haskell.org/package/base-4.19.1.0/docs/Data-IORef.html#memmodel). In other words, If you update `IORef` and `MVar` in order in one thread, the other threads may see it in a different order. This means that listeners waiting on `MVar ()` may see stale `IORef` value.

I've chosen `STM` which gives you a single view on both of variables, but you don't have `MVar` analogue for waiting in multiple threads for an update of a single `MVar`. (`TMVar` is only single-wakeup).

If we want to avoid STM here, we can define view and lock:
```haskell
, wakeLock :: !(MVar (AnyNewEpochState, SlotNo, BlockNo))
, epochStateView :: !(IORef (AnyNewEpochState, SlotNo, BlockNo))
```
assuming we won't use one value to check the other, because they can be out-of-sync.

Reply:
Why do we need to notify the listeners about changes? I could understand wanting to notify listeners of changes if the changes are short lived but they are not. What problem is this solving? Why specifically do we need this level of granularity?

Reply:
It's an alternative to `watchEpochStateView` which is polling every 100ms `EpochStateView`. Sometimes new epoch state is updated much slower, like once every few seconds. So it saves some CPU on needlessly computing the check again and again on the same input.

Reply:
But you can't implement blocking and waiting or multi-wakeup using just `IORef`.

Reply:
If you feel that `MVar`s are too much, I can rewrite this PR using polling in `watchEpochStateView`.

Reply:
> But you can't implement blocking and waiting or multi-wakeup using just IORef.

Why do we need this?

Reply:
Waiting on `MVar` is not that much more complex than polling. I'll rewrite the PR using polling then.

Reply:
The one I wrote already uses polling, you can use that one, or tune it